core/src/main/java/io/questdb/std/Decimal128.java

```diff
@@ -0,0 +1,2388 @@
+     * @return a hash code value for this object
+     */
+    @Override
+    public int hashCode() {
+        return Long.hashCode(high) ^ Long.hashCode(low) ^ Integer.hashCode(scale);
+    }
+
+    /**
+     * Check if this number is negative
+     */
+    public boolean isNegative() {
+        return high < 0;
+    }
+
+    /**
+     * Check if this number is zero
+     */
+    public boolean isZero() {
+        return high == 0 && low == 0;
+    }
+
+    /**
+     * Calculate modulo in-place.
+     *
+     * @param divisor the divisor
+     * @throws ArithmeticException if divisor is zero
+     */
+    public void modulo(Decimal128 divisor) {
+        if (divisor.isZero()) {
+            throw new ArithmeticException("Division by zero");
+        }
+
+        // Result scale should be the larger of the two scales
+        int resultScale = Math.max(this.scale, divisor.scale);
+
+        // Use simple repeated subtraction for modulo: a % b = a - (a / b) * b
+        // First compute integer division (a / b)
+        // We store this for later usage
+        long thisH = this.high;
+        long thisL = this.low;
+        int thisScale = this.scale;
+
+        this.divide(divisor, 0, RoundingMode.DOWN);
+
+        // Now compute this * divisor
+        this.multiply(divisor);
+
+        long qH = this.high;
+        long qL = this.low;
+        int qScale = this.scale;
+        // restore this as a
+        this.set(thisH, thisL, thisScale);
+        // Finally compute remainder: a - (a / b) * b
+        this.subtract(qH, qL, qScale);
+
+        // Handle scale adjustment
+        if (this.scale != resultScale) {
+            if (this.scale < resultScale) {
+                int scaleUp = resultScale - this.scale;
+                boolean isNegative = isNegative();
+                if (isNegative) {
+                    negate();
+                }
+                multiplyByPowerOf10InPlace(scaleUp);
+                if (isNegative) {
+                    negate();
+                }
+            } else {
+                int scaleDown = this.scale - resultScale;
+                for (int i = 0; i < scaleDown; i++) {
+                    divideBy10InPlace();
+                }
+            }
+            this.scale = resultScale;
+        }
+    }
+
+    /**
+     * Multiply this Decimal128 by another (in-place)
+     *
+     * @param other The Decimal128 to multiply by
+     */
+    public void multiply(Decimal128 other) {
+        multiply(other.high, other.low, other.scale);
+    }
+
+    /**
+     * Negate this number in-place
+     */
+    public void negate() {
+        // Special case: negating zero should remain zero
+        if (this.high == 0 && this.low == 0) {
+            return;
+        }
+
+        long oldLow = this.low;
+
+        // Two's complement: invert all bits and add 1
+        this.low = ~this.low + 1;
+        this.high = ~this.high;
+
+        // Check for carry from low
+        if (this.low == 0 && oldLow != 0) {
```

Comment:
this looks odd. a value overflows if and only if all `low` bits are 0s. 
No other check is needed? 

I think it can be simplified to just this:
```java
    public void negate() {
        // Two's complement: invert all bits and add 1
        this.low = ~this.low + 1;
        this.high = ~this.high;

        // Check for carry from low
        if (this.low == 0) {
            this.high += 1;
        }
    }
```
no special-case is needed. (maybe for null sentinels?)

Reply:
do you still need the special case for 0? with the new logic it should not be needed at all. 

1. initial state: all bits on both lo and hi components are zero. 
2. so you invert `lo`  bits -> it's all `1s`
3. you add `1` to `lo` -> it overflows -> `lo` is all `0s`
4. you negate `hi`  -> it's all `1s`
5. `lo` is `0`  -> you add 1 to `hi` -> hi overflow -> it's all `0s`
6. at this point all bits on both `hi` and `lo` are `0`. 
7. done

Reply:
I realized my previoius comment was ambigious, apologies for that. it's clear why you have to carry on the overflow. my question was meant to be about the special case at the top of this method:
```java
       if (this.high == 0 && this.low == 0) {
            return;
        }
```
IMHO this is redundant? since even without this branch it negates zero correctly.